### PR TITLE
Add tests for PageActionInvokerProvider

### LINKED ISSUE #5836: Tests for PageActionInvokerProvider need to be improved
These tests have some major gaps right now, importantly handlers/handler discovery aren't included.  https://github.com/aspnet/Mvc/blob/dev/test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs

We might want to relayer this code into a more sensible arrangement, like having the loader return the `CompiledPageActionDescriptor`.

### PR DESCRIPTION
**Pull request by ryanbrandenburg**
Fixes #5836. This is my first pass at adding tests for PageActionInvokerProvider. Please let me know if I've missed any important scenarios or if I'm enforcing incorrect behavior anywhere as I'm still learning about RazorPages.

### CODE REVIEW

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs [hunk 1]
```diff
@@ -70,6 +73,67 @@ public void OnProvidersExecuting_WithEmptyModel_PopulatesCacheEntry()
...
+        {
+            // Arrange
+            var descriptor = new PageActionDescriptor
+            {
+                RelativePath = "Path1",
+                FilterDescriptors = new FilterDescriptor[0],
+            };
+            Func<PageContext, object> factory = _ => null;
+            Action<PageContext, object> releaser = (_, __) => { };
+            Func<PageContext, object> modelFactory = _ => null;
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs]: Don't need these. Just use `Mock.Of<IPageModelFactoryProvider>()`

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs [hunk 2]
```diff
@@ -299,6 +362,238 @@ public void OnProvidersExecuting_UpdatesEntriesWhenActionDescriptorProviderColle
...
+
+            var descriptor = new PageActionDescriptor
+            {
+                RelativePath = "Path1",
+                FilterDescriptors = new FilterDescriptor[0],
+            };
+            var descriptorCollection1 = new ActionDescriptorCollection(new[] { descriptor }, version: 1);
+            var descriptorCollection2 = new ActionDescriptorCollection(new[] { descriptor }, version: 2);
+            var actionDescriptorProvider = new Mock<IActionDescriptorCollectionProvider>();
+            actionDescriptorProvider.SetupSequence(p => p.ActionDescriptors)
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs]: Why do you need to do this?

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs [hunk 3]
```diff
@@ -299,6 +362,238 @@ public void OnProvidersExecuting_UpdatesEntriesWhenActionDescriptorProviderColle
             Assert.NotSame(entry1, entry2);
         }
 
+        [Fact]
+        public void OnProviderExecuting_PagePopulates()
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs]: Need a better name

Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs]: Also, it doesn't look like it does any verification for the page.

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs [hunk 4]
```diff
@@ -299,6 +362,238 @@ public void OnProvidersExecuting_UpdatesEntriesWhenActionDescriptorProviderColle
...
+            invokerProvider.OnProvidersExecuting(context);
+
+            // Assert
+            var entry = Assert.Single(invokerProvider.TestCache.Entries);
+
+            Assert.Equal(2, entry.Value.ActionDescriptor.HandlerMethods.Count);
+        }
+
+        [Fact]
+        public void OnProviderExecuting_ModelPopulates()
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs]: I think this test is covered by https://github.com/aspnet/Mvc/blob/d7bb910114e9232494f6a6751646d86ec3996dd7/test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs#L187-L188

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs [hunk 5]
```diff
@@ -299,6 +362,238 @@ public void OnProvidersExecuting_UpdatesEntriesWhenActionDescriptorProviderColle
...
+            var entry = Assert.Single(invokerProvider.TestCache.Entries);
+
+            Assert.NotNull(entry.Value.ModelFactory);
+            Assert.Equal(2, entry.Value.ActionDescriptor.HandlerMethods.Count);
+        }
+        
+        [Theory]
+        [InlineData(typeof(TestSetPageWithModel))]
+        [InlineData(typeof(TestSetPageModel))]
+        public void PopulateHandlerMethodDescriptors_FindsAll(Type pageType)
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs]: `FindsAllHandlerMethodDescriptors`?

Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs]: Also, not super sure this is different from https://github.com/aspnet/Mvc/blob/d7bb910114e9232494f6a6751646d86ec3996dd7/test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs#L76

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs [hunk 6]
```diff
@@ -299,6 +362,238 @@ public void OnProvidersExecuting_UpdatesEntriesWhenActionDescriptorProviderColle
...
+                razorPageFactoryProvider: CreateRazorPageFactoryProvider(),
+                razorProject: razorProject);
+
+            var compiledDescriptor = CreateCompiledPageActionDescriptor(descriptor);
+
+            // Act
+            var factories = invokerProvider.GetPageStartFactories(compiledDescriptor);
+
+            // Assert
+            Assert.Equal(3, factories.Count);
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs]: You want to verify that the PageFactoryProvider is invoked with the right values of `_PageStart`.

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs [hunk 7]
```diff
@@ -299,6 +362,238 @@ public void OnProvidersExecuting_UpdatesEntriesWhenActionDescriptorProviderColle
...
+        }
+
+        [Fact]
+        public void GetPageStartFactories_IgnoreCasing()
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs]: This isn't true. Just a behavior of the `TestFileProvider`. Can you try and make it case sensitive and see what fails (apart from this test)?

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs [hunk 8]
```diff
@@ -299,6 +362,238 @@ public void OnProvidersExecuting_UpdatesEntriesWhenActionDescriptorProviderColle
...
+            // Act
+            var factories = invokerProvider.GetPageStartFactories(compiledDescriptor);
+
+            // Assert
+            Assert.Equal(3, factories.Count);
+        }
+
+        private IRazorPageFactoryProvider CreateRazorPageFactoryProvider()
+        {
+            Func<IRazorPage> factory1 = () => null;
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs]: Declare this inline.

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs (hunk 5 above)
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs]: What would be useful is to have tests for inheritance hierarchies, protected methods, overloads, methods overriden with with `new` etc,

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs [hunk 9]
```diff
@@ -299,6 +356,152 @@ public void OnProvidersExecuting_UpdatesEntriesWhenActionDescriptorProviderColle
...
+
+            var actionDescriptor = CreateCompiledPageActionDescriptor(descriptor, pageType);
+
+            var type = actionDescriptor.ModelTypeInfo ?? actionDescriptor.PageTypeInfo;
+
+            // Act
+            PageActionInvokerProvider.PopulateHandlerMethodDescriptors(type, actionDescriptor);
+
+            // Assert
+            Assert.Equal(expectedCount, actionDescriptor.HandlerMethods.Count);
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs]: Could you assert the `MethodInfo`? Count really doesn't verify enough.

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs (hunk 6 above)
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs]: If you do `mock.Setup(...).Verifiable()` during the setup, you can do `mock.Verify()` to verify the assumption about _PageStarts

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs [hunk 10]
```diff
@@ -299,6 +356,152 @@ public void OnProvidersExecuting_UpdatesEntriesWhenActionDescriptorProviderColle
...
+                razorPageFactoryProvider: CreateRazorPageFactoryProvider(),
+                razorProject: razorProject);
+
+            var compiledDescriptor = CreateCompiledPageActionDescriptor(descriptor);
+
+            // Act
+            var factories = invokerProvider.GetPageStartFactories(compiledDescriptor);
+
+            // Assert
+            Assert.Equal(0, factories.Count);
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs]: `Assert.Empty(factories);`

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageLoader.cs [hunk 11]
```diff
@@ -33,22 +34,20 @@ public class DefaultPageLoader : IPageLoader
             _logger = logger;
         }
 
-        public Type Load(PageActionDescriptor actionDescriptor)
+        public CompiledPageActionDescriptor Load(PageActionDescriptor actionDescriptor)
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageLoader.cs]: Can you file a follow up work item to add tests for this method + cleaning up the resx string here.

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs [hunk 12]
```diff
@@ -299,6 +356,152 @@ public void OnProvidersExecuting_UpdatesEntriesWhenActionDescriptorProviderColle
             Assert.NotSame(entry1, entry2);
         }
 
+        [Theory]
+        [InlineData(typeof(InheritsMethods), 2)]
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs]: This isn't really a good candidate for theory tests. Let's split it out and call out what we expect in each of these 3 scenarios e.g. `PopulateHandlerMethodDescriptors_DiscoversHandlersFromBaseType` etc.

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs [hunk 13]
```diff
@@ -334,6 +537,50 @@ public void OnProvidersExecuting_UpdatesEntriesWhenActionDescriptorProviderColle
...
+        }
+
+        private class ProtectedModel
+        {
+            protected void OnGet()
+            {
+
+            }
+
+            private void OnPost()
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs]: Could you also add a `OnGet<TVal>` as a scenario. We should filter those out of discovery.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageLoader.cs (hunk 11 above)
Comment (ryanbrandenburg) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageLoader.cs]: I created #5927 for this.

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs (hunk 13 above)
Comment (ryanbrandenburg) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerProviderTest.cs]: Added.

**Comment by ryanbrandenburg** [comment]
:up::date:

**Pull request merged**